sdks/python/apache_beam/runners/worker/worker_pool_main.py

```diff
@@ -0,0 +1,150 @@
+                     '--logging_endpoint=%s'
+                     % start_worker_request.logging_endpoint.url,
+                     '--artifact_endpoint=%s'
+                     % start_worker_request.artifact_endpoint.url,
+                     '--provision_endpoint=%s'
+                     % start_worker_request.provision_endpoint.url,
+                     '--control_endpoint=%s'
+                     % start_worker_request.control_endpoint.url,
+                    ]
+
+        logging.warn("Starting worker with command %s" % command)
+        worker_process = subprocess.Popen(command, stdout=subprocess.PIPE)
+        self._worker_processes[start_worker_request.worker_id] = worker_process
+
+        # Register to kill the subprocess on exit.
+        atexit.register(worker_process.kill)
+      else:
+        worker = sdk_worker.SdkHarness(
+            start_worker_request.control_endpoint.url,
+            worker_count=self._worker_threads,
+            worker_id=start_worker_request.worker_id)
+        worker_thread = threading.Thread(
+            name='run_worker_%s' % start_worker_request.worker_id,
+            target=worker.run)
+        worker_thread.daemon = True
+        worker_thread.start()
+
+      return beam_fn_api_pb2.StartWorkerResponse()
+    except Exception as exn:
+      return beam_fn_api_pb2.StartWorkerResponse(error=str(exn))
+
+  def StopWorker(self, stop_worker_request, unused_context):
+    worker_process = self._worker_processes.pop(stop_worker_request.worker_id)
+    if worker_process:
+      logging.info("Stopping worker %s" % stop_worker_request.worker_id)
+      threading.Timer(1, lambda: worker_process.kill()).start()
+      worker_process.communicate()
+    return beam_fn_api_pb2.StartWorkerResponse()
+
+
+def main(argv=None):
+  """Entry point for worker pool service for external environments."""
+
+  parser = argparse.ArgumentParser()
+  parser.add_argument('--worker_threads',
+                      type=int,
+                      default=12,
+                      help='Number of threads per SDK worker.')
+  parser.add_argument('--container_executable',
+                      type=str,
+                      default=None,
+                      help='Executable that implements the container contract.')
+  parser.add_argument('--servicer_port',
+                      type=int,
+                      required=True,
+                      help='Bind port for the worker pool service.')
+
+  args, _ = parser.parse_known_args(argv)
+
+  address, server = (BeamFnExternalWorkerPoolServicer.start(
+      worker_threads=args.worker_threads,
+      use_process=True,
+      port=args.servicer_port,
+      container_executable=args.container_executable))
+  logging.getLogger().setLevel(logging.INFO)
+  logging.info('Started worker pool servicer at port: %s with executable: %s',
+               address, args.container_executable)
+  try:
+    while True:
+      time.sleep(60 * 60 * 24)
```

Comment:
Is sleep the easiest way to allow key board interrupts to pass through? There is also `thread.Event` but I suppose this will do here.

Reply:
This seems to work fine for now. In the future there will be a better way to do it: https://github.com/grpc/grpc/pull/19299/files#diff-453fbc3ace43be70e69f9fd839788ce3R963